app/src/main/java/org/oppia/app/hintsandsolution/HintsViewModel.kt

```diff
@@ -36,25 +36,23 @@ class HintsViewModel @Inject constructor() : HintsAndSolutionItemViewModel() {
     itemList.clear()
     for (index in hintList.indices) {
       if (itemList.isEmpty()) {
-        val hintsAndSolutionViewModel = HintsViewModel()
-        hintsAndSolutionViewModel.title.set(hintList[index].hintContent.contentId)
-        hintsAndSolutionViewModel.hintsAndSolutionSummary.set(hintList[index].hintContent.html)
-        hintsAndSolutionViewModel.isHintRevealed.set(hintList[index].hintIsRevealed)
-        itemList.add(hintsAndSolutionViewModel as HintsAndSolutionItemViewModel)
-      } else if (index == itemList.size) {
-        val isLastHintRevealed = (itemList[index - 1] as HintsViewModel).isHintRevealed.get()!!
-        if (isLastHintRevealed) {
-          val hintsAndSolutionViewModel = HintsViewModel()
-          hintsAndSolutionViewModel.title.set(hintList[index].hintContent.contentId)
-          hintsAndSolutionViewModel.hintsAndSolutionSummary.set(hintList[index].hintContent.html)
-          hintsAndSolutionViewModel.isHintRevealed.set(hintList[index].hintIsRevealed)
-          itemList.add(hintsAndSolutionViewModel as HintsAndSolutionItemViewModel)
+        addHintToList(hintList[index])
+      } else {
+        val isLastHintRevealed =
+          (itemList[itemList.size - 2] as HintsViewModel).isHintRevealed.get()!!
+        if (isLastHintRevealed && index <= newAvailableHintIndex.get()!! / 2) {
+          addHintToList(hintList[index])
+        } else {
+          break
         }
       }
     }
-
-    val isLastHintRevealed = (itemList[itemList.size - 1] as HintsViewModel).isHintRevealed.get()!!
-    if (solution.hasExplanation() && hintList.size == itemList.size && isLastHintRevealed) {
+    val isLastHintRevealed = (itemList[itemList.size - 2] as HintsViewModel).isHintRevealed.get()!!
+    if (solution.hasExplanation() &&
+      hintList.size * 2 == itemList.size &&
+      isLastHintRevealed &&
+      allHintsExhausted.get()!!
+    ) {
       val solutionViewModel = SolutionViewModel()
```

Comment:
Just like `addHintToList` can we create another method for this `addSolutionToList`?
This might makes `processHintList` looks clean and separable.

Reply:
Done.